### PR TITLE
fix(vrl): update type_def after del

### LINKED ISSUE #8068: `del` doesn't remove the typedef for the deleted field
This Vrl compiles when it shouldn't. `del` currently is not removing the typedef for `.onk` and so the call to `contains` thinks it is receiving a string. It actually gets a null and thus panics.

```coffeescript
    .onk = "nork"
    del(.onk)
    if contains(.onk, "rk") {
      log("no")
    }
```

### PR DESCRIPTION
**Pull request by StephenWakely**
Closes #8068 

This ensures the type def for any paths are removed after they are removed using `del`.

To enable this fix I have had to add a new function, `update_state` to the `Expression` trait. This gives an expressions the option to mutate the state during the compilation stage. Note, I am only invoking this function whilst compiling a function call. For any other expressions, even if they implement this function it will currently have no effect. Technically, this may not be completely correct, but in the interests of not doing work when it isn't really necessary I have not pursued it further.

Signed-off-by: Stephen Wakely <fungus.humungus@gmail.com>

### CODE REVIEW

File: lib/vrl/compiler/src/type_def.rs [hunk 1]
```diff
@@ -1172,6 +1172,141 @@ impl TypeDef {
...
+                            TypeKind::Object(
+                                object
+                                    .iter()
+                                    .filter_map(|(field, kindinfo)| match field {
+                                        Field::Field(name) if name == fieldname.as_str() => None,
+                                        _ => Some((field.clone(), kindinfo.clone())),
+                                    })
+                                    .collect(),
+                            )
+                        }
```
Comment (spencergilbert) [inline lib/vrl/compiler/src/type_def.rs:1213]: Whats the difference between these two match arms?

Comment (spencergilbert) [reply lib/vrl/compiler/src/type_def.rs:1213]: I guess my questions is generally between the `thing(_)` vs `thing(specific)`

**Review by spencergilbert — APPROVED** [review]

File: lib/vrl/compiler/src/type_def.rs (hunk 1 above)
Comment (leebenson) [reply lib/vrl/compiler/src/type_def.rs:1213]: I think in the first one, the important part is matching the `Some` on the 3rd element; the field didn't need binding.

Whereas the second arm is `None` in the 3rd element, and the fieldname is needed.

Comment (spencergilbert) [reply lib/vrl/compiler/src/type_def.rs:1213]: Oh duh - I misread that and thought it was just matching on the end bit. Not enough ☕

Comment (StephenWakely) [reply lib/vrl/compiler/src/type_def.rs:1213]: It's actually wrong!

The first arm means that we aren't at the end of the path and so we need to recurse down, the second arm we are at the end of the path and need to remove the field.

The problem is it is recursing down all the the fields of the object, whereas it should only recurse down the one indicated by the current field.

Well spotted, you could see something wasn't quite right!

**Pull request merged**